service_capacity_modeling/capacity_planner.py

```diff
@@ -576,6 +576,58 @@ def _plan_certain(  # pylint: disable=too-many-positional-arguments
         num_results = num_results or self._default_num_results
         return reduce_by_family(plans)[:num_results]
 
+    # Calculates the minimum cpu, memory, and network requirements based on desires.
+    def _per_instance_requirements(self, desires) -> Tuple[int, float, float]:
+
+        # Applications often set fixed reservations of heap or OS memory
+        per_instance_mem = (
+            desires.data_shape.reserved_instance_app_mem_gib
+            + desires.data_shape.reserved_instance_system_mem_gib
+        )
+
+        # Applications often require a minimum amount of true parallelism
+        per_instance_cores = int(
+            math.ceil(
+                desires.query_pattern.estimated_read_parallelism.mid
+                + desires.query_pattern.estimated_write_parallelism.mid
+            )
+        )
+
+        per_instance_net = 0.0
+
+        current_capacity = (
+            None
+            if desires.current_clusters is None
+            else (
+                desires.current_clusters.zonal[0]
+                if len(desires.current_clusters.zonal)
+                else desires.current_clusters.regional[0]
+            )
+        )
+        # Return early if we dont have current_capacity set.
+        if current_capacity is None or current_capacity.cluster_instance is None:
+            return (per_instance_cores, per_instance_mem, per_instance_net)
+
+        # Calculate CPU requirements based on current capacity
+        current_cpu_utilization = current_capacity.cpu_utilization.high / 100.0
+        current_instance_cpus = current_capacity.cluster_instance.cpu
+        required_cores = math.ceil(current_cpu_utilization * current_instance_cpus)
+        per_instance_cores = max(per_instance_cores, required_cores)
+
+        # Calculate memory requirements based on current capacity
+        current_memory_utilization_gib = (
+            current_capacity.memory_utilization_mib.high / 1024
+        )
+        per_instance_mem = max(per_instance_mem, current_memory_utilization_gib)
+
+        # Calculate network requirements based on current capacity
+        current_network_utilization_mbps = (
+            current_capacity.network_utilization_mbps.high
+        )
+        per_instance_net = max(per_instance_net, current_network_utilization_mbps)
```

Comment:
Does per node network make sense as a filter? I get memory (true usage, not allocated) and cores, but network is not generally in-compressible. I think as long as the cluster network capacity is high enough we could assume load balancing will resolve this?

Reply:
On Titus we generally needed some per-container guaranteed network bw, but thats due to multi-tenancy. You are probably right that its enough for the cluster to have enough network.

Reply:
Removed network as a filter.